group.go

```diff
@@ -6,9 +6,19 @@ import (
 	"strings"
 )
 
+type Middleware func(next HandlerFunc) HandlerFunc
```

Comment:
With latest golang's version, there is some traction to define middlewares in router-agnostic way, ie: `func(http.Handler) http.Handler`

Maybe, helper function to wrap middleware(s) using standard http package would be useful. For example [echo#WrapMiddleware](https://godoc.org/github.com/labstack/echo#WrapMiddleware). 

Chi router even uses middleware definition based on standard http package directly [chi#Mux.Use](https://godoc.org/github.com/go-chi/chi#Mux.Use).

Existing middlewares:

- https://godoc.org/github.com/rs/cors#Cors.Handler (it's method, but can be used as `func(http.Handler) http.Handler`),
- https://godoc.org/github.com/rs/formjson#Handler,
- https://www.godoc.org/github.com/philippgille/ln-paywall/wall#NewHandlerMiddleware

Reply:
I am all for it, but passing `params` in a fast way without allocation is the problem, e.g. you could

``` go
ctx = ctx.WithValue("params", params) // wastes some CPU
req = req.WithContext(ctx) // allocates
```

It probably does not matter in practice (except in benchmarks tournament) but it is also trivial to use standard middlewares with current design too:

``` go
router.Use(func(next HandlerFunc) HandlerFunc {
	h := cors.New(...)
	return func(w http.ResponseWriter, r *http.Request, params map[string]string) {
		h.HandlerFunc(w, r)
		next(w, r, params)
	}
})
```

httptreemux can even provide an adapter so the code above becomes

``` go
h := cors.New(...)
router.Use(httptreemux.Adapter(h.HandlerFunc))
```

Reply:
Or even

``` go
h := cors.New(...)
router.UseHandlerFunc(h.HandlerFunc)
```

I will add it - good idea :)

Reply:
> I am all for it, but passing params in a fast way without allocation is the problem, e.g. you could...

Makes sense. Maybe for httptreemux, `WrapMiddleware(m func(http.Handler) http.Handler) Middleware` makes more sense.

> Or even

I meant to stick to [http.Handler](https://godoc.org/net/http#Handler), which is more widely used, as [http.HandlerFunc](https://godoc.org/net/http#HandlerFunc) is just an adapter for functions:

```
h := cors.New(...)
router.Use(httptreemux.WrapMiddleware(h.Handler))
```

Reply:
For now I've added only `UseHandlerFunc` since it is trivial and without "side effects".

It is possible to support `Handler(h http.Handler) http.Handler` too but there are 2 possibilities:
1. don't pass `params` further in chain which I guess is not a good option unless params are in `req.Context` already
2. allocate a closure for `params` on each request to preserve `params`
3. save params in `req.Context`

My preference would be 2.) and drop `ContextGroup` support (only half-joking) :)

Reply:
I would go for maximum compatibility with Go ecosystem. However, @dimfeld  should call shots.

Also, although function definition is convenient, interface based types allow to provide much more optional functionality, and not demand these extra functionalities to be implemented.

For example it is possible to list/print registered middlewares in user-friendly way, using [fmt#Stringer](https://godoc.org/fmt#Stringer): `if asStringer, ok := middleware.(fmt.Stringer); ok { ... } else { ... }`.  Or, via `type MiddlewareDescriber interface { DescribeMiddleware() string }` This could allow to produce debug output after initialization of router, and list all registered routes and used middlewares in user-friendly way.